### PR TITLE
mgr/dashboard: Replace Replica size and Erasure code profile with a single column

### PR DESCRIPTION
**Pull request by avanthakkar**
Replaces Type, Replica Size & Erasure Coded Profile columns with "Data Protection" column. Hidden Crush Rule set column and in case if any admin finds it useful to show it can still be done by removing `isHidden: true` line.

Fixes: https://tracker.ceph.com/issues/47901

Signed-off-by: Avan Thakkar <athakkar@redhat.com>

Before:
Screenshot from 2020-11-17 15-59-52

After:
Screenshot from 2020-12-09 18-42-28

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test docs

**Comment by avanthakkar** [comment]
jenkins retest this please

**Comment by avanthakkar** [comment]
jenkins retest this please

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by epuertat — COMMENTED** [review]
Looks good, but I think we could make it even more compact while still informative. What abotu something like (`replicated: xN` for replicated pools and `EC: k+m`):

image

In the above screenshot, I also hid the 'crush ruleset' column (which I think it's too low-level, and added label to the application, as in 'rbd'). The colors used are just an example, probably we should use something less eye-catching.

**Comment by avanthakkar** [comment]
> Looks good, but I think we could make it even more compact while still informative. What abotu something like (`replicated: xN` for replicated pools and `EC: k+m`):
> 
> image
> 
> In the above screenshot, I also hid the 'crush ruleset' column (which I think it's too low-level, and added label to the application, as in 'rbd'). The colors used are just an example, probably we should use something less eye-catching.

Look good to me. Currently we are not showing 'crush ruleset' details in Data Protection column, so are you suggesting to not expose it at all? Yes, regarding the colors I will try something less eye-catching. Maybe same as you have suggested for Applications label for Data Protection as well?!

**Comment by epuertat** [comment]
> > Looks good, but I think we could make it even more compact while still informative. What abotu something like (`replicated: xN` for replicated pools and `EC: k+m`):
> > image
> > In the above screenshot, I also hid the 'crush ruleset' column (which I think it's too low-level, and added label to the application, as in 'rbd'). The colors used are just an example, probably we should use something less eye-catching.
> 
> Look good to me. Currently we are not showing 'crush ruleset' details in Data Protection column, so are you suggesting to not expose it at all? Yes, regarding the colors I will try something less eye-catching. Maybe same as you have suggested for Applications label for Data Protection as well?!

@avanthakkar 👍🏼 (I was suggest to simply hide the crush ruleset, but as this is beyond this PR we may discuss this tomorrow).

**Review by epuertat — APPROVED** [review]
LGTM @avanthakkar! Thank you. Just some minor fixes and this caching issue, that we can defer to another PR (but we should start tackling asap).

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts [hunk 1]
```diff
@@ -232,6 +234,19 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     };
   }
 
+  fetchErasureCodeProfile(erasureCodeProfile: string) {
+    this.ecpService.list().subscribe((ecProfileList: ErasureCodeProfile[]) => {
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:238]: I think caching this in the front-end is a good idea. Additionally we should add a comment to the API endpoint that this is an expensive operation and the consumer of this endpoint should consider caching this. I don't know if it is a good idea to cache this on the back-end because as a consumer i expect the lastest data when calling the endpoint. But maybe this is a personal point of view. If it is a common approach to cache such things on the back-end side i'm also fine with this.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts [hunk 2]
```diff
@@ -232,6 +234,19 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
...
+      _.forEach(ecProfileList, (ecpKey) => {
+        if (ecpKey['name'] === erasureCodeProfile) {
+          this.k = ecpKey['k'];
+          this.m = ecpKey['m'];
+        }
+      });
+    });
+
+    return `EC: ${this.k}+${this.m}`;
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:247]: It's luck that this works somehow. You're formatting a string with values that are not assigned at the current state because `this.ecpService.list().subscribe` is executed AFTER this code line. You're mixing synchronous and asynchronous code here.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:247]: Yes  I see the problem. When the first it subscribes it shows 'undefined+undefined'. So how can I avoid this? Any suggestions?

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:247]: Will this work.. I just not return the string and use k & m values where the function is called (in transformsPoolData())?

Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:247]: You better use a pipe for the `Data protection` column. This will make the asynchronicity of showing this data much more easier.

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:247]: async pipe?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:238]: Perhaps it's a bad trait of mine, but I tend to prefer 'defensive coding' 🙈 It's not about mistrust, bad intentions or incompetence, but because in heavily layered (and encapsulated) systems like the dashboard, when you are coding at the top of the stack, you shouldn't care too much about the performance at the bottom of it. In the end, nobody knows better about the internals of component than the component itself...

That doesn't mean that a back-end component has to cheat its consumers: if we provide a cached response we could deliver it along with the timestamp or a uuid of the response, so the consumer can know how fresh/stale the response is.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts]: I tried with aysnc pipe but the issue is how to use it in template for `Data Protection` column? I looked if somehow async pipe can be used as column property but there doesn't seems to be one. So for now, I have removed the return statement and instead just used directly the updated values of k and m.

**Comment by avanthakkar** [comment]
jenkins retest this please

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by s0nea** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts]: @avanthakkar @votdev @epuertat IMHO we can solve 2 problems here:
1) No need to call this.ecpService.list() several times inside a loop: 1 is enough.
2) Async code issue can be resolved using ```mergeMap``` operator to ensure that we already have ecProfileList info when processing pool data.

The refactor can be something like this:
```
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
index 171d6871cce..f1c6c2c3d48 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
@@ -1,7 +1,7 @@
 import { Component, OnInit, TemplateRef, ViewChild } from '@angular/core';
 
 import _ from 'lodash';
-import { Observable } from 'rxjs';
+import { mergeMap } from 'rxjs/operators';
 
 import { PgCategoryService } from '~/app/ceph/shared/pg-category.service';
 import { ConfigurationService } from '~/app/shared/api/configuration.service';
@@ -60,9 +60,7 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
   tableStatus = new TableStatusViewCache();
   cacheTiers: any[] = [];
   monAllowPoolDelete = false;
-  k: number;
-  m: number;
-  ecProfileList$: Observable<ErasureCodeProfile[]>;
+  ecProfileList: ErasureCodeProfile[];
 
   constructor(
     private poolService: PoolService,
@@ -196,7 +194,13 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     ];
 
     this.taskListService.init(
-      () => this.poolService.getList(),
+      () =>
+        this.ecpService.list().pipe(
+          mergeMap((ecProfileList: ErasureCodeProfile[]) => {
+            this.ecProfileList = ecProfileList;
+            return this.poolService.getList();
+          })
+        ),
       undefined,
       (pools) => {
         this.pools = this.transformPoolsData(pools);
@@ -236,15 +240,15 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     };
   }
 
-  fetchErasureCodeProfile(erasureCodeProfile: string) {
-    this.ecpService.list().subscribe((ecProfileList: ErasureCodeProfile[]) => {
-      _.forEach(ecProfileList, (ecpKey) => {
-        if (ecpKey['name'] === erasureCodeProfile) {
-          this.k = ecpKey['k'];
-          this.m = ecpKey['m'];
-        }
-      });
+  getErasureCodeProfile(erasureCodeProfile: string) {
+    let ecpInfo = '';
+    _.forEach(this.ecProfileList, (ecpKey) => {
+      if (ecpKey['name'] === erasureCodeProfile) {
+        ecpInfo = `EC: ${ecpKey['k']}+${ecpKey['m']}`;
+      }
     });
+
+    return ecpInfo;
   }
 
   transformPoolsData(pools: any) {
@@ -283,8 +287,7 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
 
       if (pool['type'] === 'erasure') {
         const erasureCodeProfile = pool['erasure_code_profile'];
-        this.fetchErasureCodeProfile(erasureCodeProfile);
-        pool['data_protection'] = `EC: ${this.k}+${this.m}`;
+        pool['data_protection'] = this.getErasureCodeProfile(erasureCodeProfile);
       }
       if (pool['type'] === 'replicated') {
         pool['data_protection'] = `replica: ×${pool['size']}`;
```

**Review by nizamial09 — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts]: Looks good to me..tested it and works perfect! Thanks @alfonsomthd 
Just one thing can't we use switchMap too here? As all we need is just the latest value, but I think it doesn't make diference as we are gonna do just 1 call , right?

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts]: @avanthakkar for this particular case it does not make a difference as httpClient is just returning 1 value.
```switchMap``` is usually used when you also need cancellation logic, which is not the case.

**Review by alfonsomthd — APPROVED** [review]

**Comment by avanthakkar** [comment]
Thanks @alfonsomthd for taking care of test suites:)

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by pcuzner** [comment]
@epuertat lgtm. I like the color code for the protection type, and removing the crushrule set also makes sense - since it's visible anyway in the pool details

**Comment by nizamial09** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts [hunk 3]
```diff
@@ -120,60 +125,57 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
         cellTransformation: CellTemplate.executing
       },
       {
-        prop: 'type',
-        name: $localize`Type`,
+        prop: 'data_protection',
+        name: $localize`Data Protection`,
+        cellTransformation: CellTemplate.badge,
+        customTemplateConfig: {
+          class: 'badge-hdd'
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:132]: Wouldn't it make sense to introduce a new CSS class here? I guess you simply want to use a color here, but using a class that is not related to this column feels wrong. Maybe simply add
```
.badge-background-gray,
.badge-hdd {
  background-color: $gray-600;
  color: $white;
}
```
```suggestion
          class: 'badge-background-gray'
```
I think this is a better approach in terms of branding, too.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts [hunk 4]
```diff
@@ -120,60 +125,57 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
...
+        },
         flexGrow: 2
       },
       {
         prop: 'application_metadata',
         name: $localize`Applications`,
-        flexGrow: 3
+        cellTransformation: CellTemplate.badge,
+        customTemplateConfig: {
+          class: 'badge-ssd'
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts:141]: Wouldn't it make sense to introduce a new CSS class here? I guess you simply want to use a color here, but using a class that is not related to this column feels wrong. Maybe simply add
```
.badge-background-primary,
.badge-ssd {
  background-color: $primary;
  color: $white;
}
```
```suggestion
          class: 'badge-background-primary'
```
I think this is a better approach in terms of branding, too.

**Review by votdev — APPROVED** [review]
LGTM except my two comments.

**Comment by avanthakkar** [comment]
jenkins test dashboard
